docs/_docs/30-fasterkv-manual-locking.md

```diff
@@ -0,0 +1,266 @@
+    - The Invalid entries are finally removed during ReadCacheEvict:
+      - CAS the RC record to be removed to be Sealed. This will cause any other operations to retry.
+      - CAS the preceding RC record to point to the to-be-removed RC record's .PreviousAddress (standard singly-linked-list operations)
+      - CAS the now-removed RC record to be Invalid.
+
+The above covers single-record operations on the RC prefix. Two-record operations occur when we must outsplice one record and insplice another, because the value for a record in the RC prefix is updated, e.g. Upsert updating a record in the ReadOnly region or RMW doing a CopyUpdater (of mutable or readonly), or either of these operating updating a key that is in the RC prefix chain. The considerations here are:
+- Updating an RC record:
+  - Mark the RC record as Sealed
+  - Do the insplice as described above
+  - If the insplice succeeds, mark the RC entry as Invalid, else remove the Sealed marking and RETRY_NOW
+- Updating an in-memory main log record:
+  - Mark the main log record as Sealed
+  - Do the insplice as described above
+  - If the insplice succeeds, leave the main log entry as Sealed, else remove the Sealed marking and RETRY_NOW
+
+#### Main Log Evictions
+
+When main log pages are evicted due to memory limits, *if* there are any active `ManualFasterOperations` sessions, then each record on those pages must be examined and any locks transferred to `LockTable` entries.
+
+Transfers to the `LockTable` due to main log evictions are handled in the following manner:
+- A new `TentativeHeadAddress` (THA) field is added next to `HeadAddress`.
+- Shifting HeadAddress is now done in two steps: Update THA, then update HeadAddress
+- The purpose of THA is:
+  - No record at address < THA shall be touched for Lock operations (lock or unlock).
+  - Existing locks will remain there until transferred to the lock table.
+  - A thread that attempts to unlock a record < THA must wait for the lock table to get populated (details below)
+    - To mitigate unlocking records that were never locked, this must abandon the wait when the record to be unlocked becomes < HeadAddress.
+
+Here is the sequence of operations to perform record eviction with `LockTable` Transfer:
+- In (PageAligned)ShiftHeadAddress:
+  - TentativeHeadAddress = desired new HeadAddress
+  - BumpCurrentEpoch(() => OnPagesReadyToTransfer())
+    - OnPagesReadyToTransfer() is a new routine:
+      - Scan from OldTentativeHeadAddress to new TentativeHeadAddress:
+        - Transfer records to `LockTable`
+        - This is safe because we are executing in the epoch Drain thread
+      - OldTentativeHeadAddress = TentativeHeadAddress
+      - HeadAddress = TentativeHeadAddress
```

Comment:
Actually, here we have to do whatever ShiftHeadAddress does today, not just set HeadAddress.